be/src/exec/spill/executor.h

```diff
@@ -90,32 +91,51 @@ struct ResourceMemTrackerGuard {
     mutable MemTracker* old_tracker = nullptr;
 };
 
+struct SpillIOTaskContext {
+    bool use_local_io_executor = true;
+};
+using SpillIOTaskContextPtr = std::shared_ptr<SpillIOTaskContext>;
+
 struct IOTaskExecutor {
-    workgroup::ScanExecutor* pool;
+    workgroup::ScanExecutor* local_io_executor;
+    workgroup::ScanExecutor* remote_io_executor;
     workgroup::WorkGroupPtr wg;
 
-    IOTaskExecutor(workgroup::ScanExecutor* pool_, workgroup::WorkGroupPtr wg_) : pool(pool_), wg(std::move(wg_)) {}
+    IOTaskExecutor(workgroup::ScanExecutor* local_io_executor, workgroup::ScanExecutor* remote_io_executor,
+                   workgroup::WorkGroupPtr wg)
+            : local_io_executor(local_io_executor), remote_io_executor(remote_io_executor), wg(std::move(wg)) {}
 
-    template <class Func>
-    Status submit(Func&& func) {
-        workgroup::ScanTask task(wg.get(), func);
+    Status submit(workgroup::ScanTask task) {
+        const auto& task_ctx = task.get_work_context();
+        bool use_local_io_executor = true;
+        if (task_ctx.task_context_data.has_value()) {
+            auto io_ctx = std::any_cast<SpillIOTaskContextPtr>(task_ctx.task_context_data);
+            use_local_io_executor = io_ctx->use_local_io_executor;
+        }
+        auto* pool = use_local_io_executor ? local_io_executor : remote_io_executor;
         if (pool->submit(std::move(task))) {
             return Status::OK();
         } else {
             return Status::InternalError("offer task failed");
         }
     }
+    void force_submit(workgroup::ScanTask task) {
+        const auto& task_ctx = task.get_work_context();
+        auto io_ctx = std::any_cast<SpillIOTaskContextPtr>(task_ctx.task_context_data);
+        auto* pool = io_ctx->use_local_io_executor ? local_io_executor : remote_io_executor;
+        pool->force_submit(std::move(task));
+    }
 };
 
 struct SyncTaskExecutor {
-    template <class Func>
-    Status submit(Func&& func) {
-        workgroup::YieldContext yield_ctx;
+    Status submit(workgroup::ScanTask task) {
         do {
-            std::forward<Func>(func)(yield_ctx);
-        } while (!yield_ctx.is_finished());
+            task.run();
```

Comment:
what if run() encounter errors and is cancelled, can it return, avoiding blocking?

Reply:
> what if run() encounter errors and is cancelled, can it return, avoiding blocking?

by design, `task.run` will be invoked in thread pool, it won't return a Status. if this task need to exit, it must be processed internally by invoking ` yield_ctx.set_finished()`, then we can determine whether to run this task by checking the return value of `yield_ctx.is_finished`.